### PR TITLE
Java Kubernetes Test

### PR DESCRIPTION
**Pull request by daniel-sanche**
Added environment tests for java using the Kubernetes environment. You should now be able to run `envctl java kubernetes deploy`,  `envctl java kubernetes verify`, and  `envctl java kubernetes destroy` to manage a test cluster. You should also be able to run `nox --session "tests(language='java', platform='kubernetes')"` to basic run unit tests against the deployed code

### CODE REVIEW

File: deployable/java/src/main/java/envtest/deployable/Snippets.java
Comment (daniel-sanche) [inline deployable/java/src/main/java/envtest/deployable/Snippets.java:32]: @simonz130 is there a simpler way to write a log using the base library? Specifically with resource detection

File: deployable/java/pom.xml
Comment (simonz130) [inline deployable/java/pom.xml:29]: why do we need this if we use docker file?

File: deployable/java/pom.xml
Comment (simonz130) [inline deployable/java/pom.xml:33]: why do we need spring? Can't we spin up a simple http web server for the purposes of the test?

File: deployable/java/pom.xml
Comment (simonz130) [inline deployable/java/pom.xml:53]: why do we need springboot?

File: deployable/java/pom.xml
Comment (simonz130) [inline deployable/java/pom.xml:87]: can we reduce the # of dependencies we don't have to have to run a simple logging test?

File: deployable/java/src/main/java/envtest/deployable/DeployableApplication.java
Comment (simonz130) [inline deployable/java/src/main/java/envtest/deployable/DeployableApplication.java:106]: This is hacky and outright dangerous. Instead of using reflection to dynamically invoke methods, I suggest we do it "old school" by checking supported messages and triggering the one we want based on message value.

File: deployable/java/src/main/java/envtest/deployable/DeployableApplication.java
Comment (simonz130) [inline deployable/java/src/main/java/envtest/deployable/DeployableApplication.java:135]: nit: remove space

File: deployable/java/src/main/java/envtest/deployable/DeployableApplication.java
Comment (simonz130) [inline deployable/java/src/main/java/envtest/deployable/DeployableApplication.java:146]: I think there is a bug here: if these run, upon completion the method ends and we continue to GCR/GAE part and happen to fail only because env var is likely not defined. 
It should be safer and more clear to "return" after line 146.

File: envctl/env_scripts/java/kubernetes.sh
Comment (simonz130) [inline envctl/env_scripts/java/kubernetes.sh:54]: I wonder if it wouldn't be better to have a cluster pre-created for tests and only create a dedicated namespace and deployments in it for each test run?

File: envctl/env_scripts/java/kubernetes.sh
Comment (simonz130) [inline envctl/env_scripts/java/kubernetes.sh:61]: does it add anything on top of default scopes?

File: envctl/env_scripts/java/kubernetes.sh
Comment (simonz130) [inline envctl/env_scripts/java/kubernetes.sh:77]: what if two people run this test in parallel?

File: envctl/env_scripts/java/local.sh
Comment (simonz130) [inline envctl/env_scripts/java/local.sh:50]: should we make the topic named as the envid instead of service_name to allow parallel runs of the same test?

File: envctl/envctl
Comment (simonz130) [inline envctl/envctl:9]: don't we need uuid to add randomness into the fold?

**Review by simonz130 — CHANGES_REQUESTED** [review]

File: envctl/envctl
Comment (daniel-sanche) [reply envctl/envctl:9]: No, we want it to be consistent per-host by default - every time I run a test on my laptop, I want it to test against existing resources when possible, instead of spinning up a new cluster

The CI scripts will overwrite this value with a uuid to make sure they stay random though

File: envctl/env_scripts/java/local.sh
Comment (daniel-sanche) [reply envctl/env_scripts/java/local.sh:50]: Service name uses env_id, but adds a service-specific prefix so we can run different tests in parallel

File: envctl/env_scripts/java/kubernetes.sh
Comment (daniel-sanche) [reply envctl/env_scripts/java/kubernetes.sh:77]: They will have different env_ids, which is part of SERVICE_NAME. And presumably they will be using different test projects

File: envctl/env_scripts/java/kubernetes.sh
Comment (daniel-sanche) [reply envctl/env_scripts/java/kubernetes.sh:61]: I don't think it needs anything other than pubsub

File: envctl/env_scripts/java/kubernetes.sh
Comment (daniel-sanche) [reply envctl/env_scripts/java/kubernetes.sh:54]: That would make the test faster, but the downside is it would be another thing we'd need to manually maintain, and the ongoing resources would be much more expensive when we only really need it live for 5 mins a day

The environment tests run in parallel, and the GKE test takes around the same time as App Engine Standard (~10 mins), so I don't think speeding it up should be too much of a priority

File: deployable/java/pom.xml
Comment (daniel-sanche) [reply deployable/java/pom.xml:33]: I'm not too familiar with Java, so I added Spring since that's what we've used in the past. Let me know if you have other suggestions

File: deployable/java/src/main/java/envtest/deployable/DeployableApplication.java
Comment (daniel-sanche) [reply deployable/java/src/main/java/envtest/deployable/DeployableApplication.java:106]: Too dangerous to use even for test scaffolding? My impression is that the code here will only allow execution of functions from the "Snippets" class that fit the expected signature, which seems to be what we want. Are there other security issues I'm overlooking?

This is an important point, because we've talking about generalizing this code to potentially support other use cases, and requiring hard-coded function calls rather than dynamic ones would be a huge roadblock to that.

File: deployable/java/pom.xml
Comment (daniel-sanche) [reply deployable/java/pom.xml:87]: We likely can. I'll have to do some experimenting to see which are needed, since I'm not too familiar with this stack

File: deployable/java/pom.xml
Comment (daniel-sanche) [reply deployable/java/pom.xml:29]: goot catch, this can probably be removed

File: deployable/java/pom.xml
Comment (daniel-sanche) [reply deployable/java/pom.xml:53]: We need an HTTP server for Cloud Run, so I used the springboot architecture from bank of anthos. I'm fine with replacing this with another alternative

File: deployable/java/pom.xml
Comment (simonz130) [reply deployable/java/pom.xml:98]: I posted a sample code in another comment with minimal web server. If that doesn't work in a few min. - let's start with springboot and remove it later on - I don't want to block this PR on things that are not core.

File: deployable/java/src/main/java/envtest/deployable/Snippets.java
Comment (simonz130) [reply deployable/java/src/main/java/envtest/deployable/Snippets.java:46]: No, this is the way

File: envctl/env_scripts/java/kubernetes.sh
Comment (simonz130) [reply envctl/env_scripts/java/kubernetes.sh:54]: In cloud-code-samples we had pre-created gke cluster and each test run created a dedicated namespace on GKE and all the test objects belonged go that. Then there was a periodic task that was deleting all the namespaces that weren't deleted automatically by the test runs.
For now we can keep this approach in order to unblock the PR. However going forward we might reconsider - depending on how stable is provisioning before the actual tests.

File: envctl/envctl
Comment (simonz130) [reply envctl/envctl:9]: From framework design perspective, I think the randomness should be handled by envctl instead of by Cis using it. This makes sure the consumers of the framework don't need to make hidden assumptions about how to run the tool properly.

**Review by simonz130 — APPROVED** [review]

File: envctl/envctl
Comment (daniel-sanche) [reply envctl/envctl:9]: The problem is we *don't* want randomness during regular development. When I'm adding a test, I want it to re-use standard test cluster each time, rather than creating a new one for each test run. CI is really the special case where we really do want it to be unique for each run.

We could have it random by default and instruct contributors users to override it with a static value for development, but that's a lot more overhead than setting it to a random value within the CI scripts

**Pull request merged**
